### PR TITLE
[doc] measuring performance docs

### PR DESCRIPTION
**Pull request by bimbashrestha**
https://github.com/facebook/zstd/issues/1990

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -126,6 +126,210 @@ just `contrib/largeNbDicts` and nothing else, you can run:
...
+When measuring optimizations, something worth trying is to make sure the handware
+counters you expect to be impacted by your change are in fact being so. For example,
+if you expect the L1 cache misses to decrease with your change, you can look at the
+counter `L1-dcache-load-misses`
+* Perf hardware counters will not work on a virtual machine.
+
+#### Visual Studio
+
+TODO
+
```
Comment (bimbashrestha) [inline CONTRIBUTING.md:331]: My windows vm is being strange. I'll add this in a separate pr probably

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -126,6 +126,210 @@ just `contrib/largeNbDicts` and nothing else, you can run:
 scan-build make -C contrib/largeNbDicts largeNbDicts
 ```
 
+## Performance
+Performance is extermely important for zstd and we only merge pull requests whose performance
```
Comment (Cyan4973) [inline CONTRIBUTING.md:130]: nit: extermely -> extremely

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -126,6 +126,210 @@ just `contrib/largeNbDicts` and nothing else, you can run:
...
+impact performance takes a very long time for us to properly review. That being said, we
+always welcome contributions to improve performance (or worsen performance for the trade-off of
+something else). Please keep the following in mind before submitting a performance related PR:
+
+1. Zstd isn't as old as gzip but it has been around for time now and its evolution is
+very well documented via past Github issues and pull requests. It may be the case that your
+particular performance optimization has already been considered in the past. Please take some
+time to search through old issues and pull requests using keywords specific to your
+would-be PR. Of course, just because a topic has already been discussed (and perhaps rejected
+on some grounds) in the past, doesn't mean it isn't worth brining up again. But even in that case,
```
Comment (Cyan4973) [inline CONTRIBUTING.md:142]: nit: brining -> bringing

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -126,6 +126,210 @@ just `contrib/largeNbDicts` and nothing else, you can run:
...
+is a good place to start.
+
+### Stability
+Unfortunately, the most important aspect in being able to benchmark reliably is to have a stable
+benchmarking machine. A virtual machine, a machine with shared resources, or your laptop
+will typically not be stable enough to obtain reliable benchmark results. If you can get your
+hands on a desktop, this is usually a better scenario.
+
+Of course, benchmarking can be done on non-hyper-stable machines as well. You will just have to
+do a little more work to ensure that you are in fact measuring the changes you've made not and
```
Comment (Cyan4973) [inline CONTRIBUTING.md:169]: nit : I presume you meant "the changes you've made and not some noise" ?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -126,6 +126,210 @@ just `contrib/largeNbDicts` and nothing else, you can run:
...
+nosie. Here are some simple things you can do to make your benchmarks more stable:
+
+1. The most simple thing you can do to drastically improve the stability of your benchmark is
+to run it mulitple times and then aggregate the results of those runs. As a general rule of
```
Comment (Cyan4973) [inline CONTRIBUTING.md:173]: nit: mulitple -> multiple

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -126,6 +126,210 @@ just `contrib/largeNbDicts` and nothing else, you can run:
...
+thumb, the smaller the change you are trying to measure, the more samples of benchmark runs
+you will have to aggregate over to get reliable results. Here are some additonal things to keep in
+mind when running multiple trials:
+    * How you aggregate your samples are important. You might be tempted to use the mean of your
+    results. While this is certainly going to be a more stable number than a raw single sample
+    benchmark number, you might have more luck by taking the median. The mean is not robust to
+    outliers wheas the median is. Better still, you could simply take the fastest speed your
```
Comment (Cyan4973) [inline CONTRIBUTING.md:180]: nit: wheas -> whereas

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -126,6 +126,210 @@ just `contrib/largeNbDicts` and nothing else, you can run:
...
+    your improved stability by looking at the size of your confidence intervals as you
+    increase your sample count. These should get smaller and smaller. Eventually hopefully
+    smaller than the performance win you are expecting.
+    * Most processors will take some time to get `hot` when running anything. The observations
+    you collect during that time period will very different from the true performance number. Having
+    a very large number of sample will help alleviate this problem slightly but you can also
+    address is directly by simply not including the first `n` iterations of your benchmark in
+    your aggregations. You can determine `n` by simply looking at the results from each iteration
+    and then hand picking a good threshold after which the variance in results seems to stabilize.
+2. You cannot really get reliable benchmarks if your host machine is simultaniously running
```
Comment (Cyan4973) [inline CONTRIBUTING.md:194]: nit : simultaniously -> simultaneously

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -126,6 +126,210 @@ just `contrib/largeNbDicts` and nothing else, you can run:
...
+    on your OS:
+        * On linux boxes, you have use https://github.com/lpechacek/cpuset.
+        * On Windows, you can "Set Processor Affinity" using https://www.thewindowsclub.com/processor-affinity-windows
+        * On Mac, you can try to use their dedicated affinity API https://developer.apple.com/library/archive/releasenotes/Performance/RN-AffinityAPI/#//apple_ref/doc/uid/TP40006635-CH1-DontLinkElementID_2
+3. To benchmark, you will likely end up writing a seperate c/c++ program that will link libzstd.
+Dynamically linking your library will introduce some added variation (not a large amount but
+definitly some). Statically linking libzstd will be more stable. Static libraries should
+be enabled by default when building zstd.
+
+Those are the low hanging fruit. You can also try the following more advanced techniques
```
Comment (Cyan4973) [inline CONTRIBUTING.md:211]: opinion : I believe these are not "advanced" techniques, they are actually very important and useful.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -126,6 +126,210 @@ just `contrib/largeNbDicts` and nothing else, you can run:
...
+tmpfs avoids touching a real storage system, which can have a pretty big variability.
+
+Also check our LLVM's guide on benchmarking here: https://llvm.org/docs/Benchmarking.html
+
+### zstdbench
+The fastest signal you can get regarding your performance changes is via the in-build zstd cli
+bench option. You can run Zstd as you typically would for your scenario using some set of options
+and thn additinally also specify the `-b#` option. Doing this will run our benchmarking pipeline
+for that options you have just provided. If you want to look at the internals of how this
+benchmarking script works, you can check out benchzstd.c file under tests.
```
Comment (Cyan4973) [inline CONTRIBUTING.md:227]: > benchzstd.c file under tests

minor : prefer `programs/benchzstd.c`

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -126,6 +126,210 @@ just `contrib/largeNbDicts` and nothing else, you can run:
...
+$ make && cp zstd zstd-old
+$ git checkout <commit-after-your-change>
+$ make && cp zstd zstd-new
+$ zstd-old -i5 -b1 <your-test-data>
+ 1<your-test-data>         :      8990 ->      3992 (2.252), 302.6 MB/s , 626.4 MB/s
+$ zstd-new -i5 -b1 <your-test-data>
+ 1<your-test-data>         :      8990 ->      3992 (2.252), 302.8 MB/s , 628.4 MB/s
+```
+
+Unless your performace win is large enough to be visibile despite the intrinsic noise
```
Comment (Cyan4973) [inline CONTRIBUTING.md:246]: nit: visibile -> visible

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -126,6 +126,210 @@ just `contrib/largeNbDicts` and nothing else, you can run:
...
+There are a number of great profilers out there. We're going to briefly mention how you can
+profile your code using `instruments` on mac, `perf` on linux and `visual studio profiler`
+on windows.
+
+Say you have an idea for a change that you think will provide some good performance gains
+for level 1 compression on Zstd. Typically this means, you have identified a section of
+code that you think can be made to run faster.
+
+The first thing you will want to do is make sure that the piece of code is actually taking up
+a notable amount of time to run. It is usually not worth optimzing something which accounts for less than
```
Comment (Cyan4973) [inline CONTRIBUTING.md:263]: nit: optimzing -> optimizing

**Comment by Cyan4973** [comment]
These are generally pretty good advises.
At least, it invites caution before claiming victory from a single benchmark in a single scenario.
Beyond nit, my only comment is that some advises classified "advanced" are actually fairly good and important, and seem adequate even for benchmark novices.

**Comment by bimbashrestha** [comment]
> Beyond nit, my only comment is that some advises classified "advanced" are actually fairly gd and important, and seem adequate even for benchmark novices.

Sure. I'll just make that one section instead of splitting it into two

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -203,28 +203,24 @@ on Windows. You will get more stable benchmark results of you end those processe
...
-2. Disable frequency scaling, turbo boost and address space randomization (this will vary by OS)
-3. Try to avoid storage. On some systems you can use tmpfs. Putting the program, inputs and outputs on
+5. Disable frequency scaling, turbo boost and address space randomization (this will vary by OS)
+6. Try to avoid storage. On some systems you can use tmpfs. Putting the program, inputs and outputs on
 tmpfs avoids touching a real storage system, which can have a pretty big variability.
 
 Also check our LLVM's guide on benchmarking here: https://llvm.org/docs/Benchmarking.html
 
-### zstdbench
+### Becnhzstd
```
Comment (Cyan4973) [inline CONTRIBUTING.md:218]: nit: `benchzstd`
or better yet : `zstd` benchmark
(not sure if it's important for this paragraph to be titled with the name of a source file)

**Pull request merged**